flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/hooks/MasterHooks.java

```diff
@@ -267,6 +269,112 @@ else if (!allowUnmatchedState) {
 	}
 
 	// ------------------------------------------------------------------------
+	//  hook management
+	// ------------------------------------------------------------------------
+
+	/**
+	 * Wraps a hook such that the user-code classloader is applied when the hook is invoked.
+	 * @param hook the hook to wrap
+	 * @param userClassLoader the classloader to use
+	 */
+	public static <T> MasterTriggerRestoreHook<T> wrapHook(MasterTriggerRestoreHook<T> hook, ClassLoader userClassLoader) {
+		return new WrappedMasterHook<T>(hook, userClassLoader);
+	}
+
+	@VisibleForTesting
+	static class WrappedMasterHook<T> implements MasterTriggerRestoreHook<T> {
+
+		private final MasterTriggerRestoreHook<T> hook;
+		private final ClassLoader userClassLoader;
+
+		WrappedMasterHook(MasterTriggerRestoreHook<T> hook, ClassLoader userClassLoader) {
+			this.hook = hook;
+			this.userClassLoader = userClassLoader;
+		}
+
+		@Override
+		public String getIdentifier() {
+			Thread thread = Thread.currentThread();
+			ClassLoader originalClassLoader = thread.getContextClassLoader();
+			thread.setContextClassLoader(userClassLoader);
```

Comment:
I disagee with that approach due to lack of symmetry.    The try..finally idiom is typically written as: